---
title: "2_DESeq_analysis"
output: html_document
date: "`r Sys.Date()`"
---

```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE, warning = TRUE, message = FALSE, cache = TRUE)
```

```{bash, engine.opts='-l'}
echo $PATH
```

### Load libraries
```{r load_libraries, inlcude = TRUE}

## Install Rtools directly from (https://cran.r-project.org/bin/windows/Rtools/), then install these on first run:
install.packages("BiocManager")
BiocManager::install("DESeq2")
BiocManager::install("vsn")
BiocManager::install("tidybulk")
BiocManager::install("goseq")
BiocManager::install("affycoretools")
BiocManager::install("EnhancedVolcano")
BiocManager::install("pcaExplorer")
BiocManager::install("apeglm")
BiocManager::install("PCAtools")


# List of packages we want to install (run every time)
load.lib<-c("DESeq2","edgeR","goseq","dplyr","GenomicFeatures","data.table","calibrate","affycoretools","data.table","vsn","tidybulk","ggplot2","cowplot","pheatmap","gplots","RColorBrewer","EnhancedVolcano","pcaExplorer","readxl","apeglm","ashr","tibble","plotly","sqldf","PCAtools","ggpubr")

# Select only the packages that aren't currently installed (run every time)
install.lib <- load.lib[!load.lib %in% installed.packages()]

# And finally we install the missing packages, including their dependency.
for(lib in install.lib) install.packages(lib,dependencies=TRUE)
# After the installation process completes, we load all packages.
sapply(load.lib,require,character=TRUE)
                        
```

#Set ggplot theme
```{r ggplot_theme, include=FALSE}

my_theme <- theme(line              = element_line(size=1.5),
                  rect              = element_rect(size=1.5),
                  text              = element_text(size=14,color="black"),
                  panel.background  = element_blank(),
                  panel.grid.major  = element_blank(), 
                  panel.grid.minor  = element_blank(),
                  axis.text.x       = element_text(size=16,color="black"),
                  axis.text.y       = element_text(size=16,color="black"),
                  axis.title.x      = element_text(margin = margin(t = 10, r = 0, b = 0, l = 0)),
                  axis.title.y      = element_text(margin = margin(t = 0, r = 10, b = 0, l = 0)),
                  axis.ticks.x      = element_line(color="black"),
                  axis.ticks.y      = element_line(color="black"),
                  # axis.line         = element_line(color = "black", size = 0.1),
                  panel.border      = element_rect(color = "black", fill=NA, size=1.5),
                  legend.key        = element_blank()) # removes background of legend bullets
```

#Load data
```{r load_data, inlcude = TRUE}

# Load LOCID to gene description tables, compare, find gigas and mito genes
a1 <- read.delim("../data/GCF_902806645.1_cgigas_uk_roslin_v1_feature_table.txt", header = TRUE)
a1 <- a1 %>% filter(X..feature == "CDS")
a1 <- a1[,c("symbol","name")]
colnames(a1) <- c("LOCID","description")
a1 <- distinct(a1, LOCID, .keep_all = TRUE)
a2 <- read.delim("../data/LOCID_38222_Roslin_Gigas_mito.txt", header = TRUE)
colnames(a2) <- c("LOCID","description")
a3 <- anti_join(a2,a1, by="LOCID")
a1$feature <- 'genome'
a3$feature <- 'mito'
LOCID_Roslin_gigas_gene_mito_table <- as.data.frame(rbind(a1,a3))
LOCID_Roslin_gigas_gene_mito_table <- distinct(LOCID_Roslin_gigas_gene_mito_table, LOCID, .keep_all = TRUE)
write.table(LOCID_Roslin_gigas_gene_mito_table, "../data/LOCID_Roslin_gigas_gene_mito.txt")
rm(a1,a2,a3)

# Import experimental design information
trt_list <- read.csv("../data/treatments.csv", sep=',', header=TRUE, row.names = "ID")
coldata <- trt_list[,c("ploidy","condition")]

# Import gene_count_matrix
cts <- as.matrix(read.csv("../data/20220608_gene_count_matrix.csv", sep=',', header=TRUE, row.names = "gene_id"))
colnames(cts) <- row.names(trt_list) #!make sure column and row names for coldata and cts match
all(colnames(cts) %in% rownames(coldata))

# Factor independent variables
coldata$ploidy <- factor(coldata$ploidy)
coldata$condition <- factor(coldata$condition)
coldata$group <- factor(trt_list$group) # for individual pair-wise comparisons

# Remove bad samples
# MuliQC report: N56, X44
# Pheatmap: M43, N57, R53, R62, T62, X42, X44 
coldata <- coldata[!(row.names(coldata) %in% c('N56', 'X44', 'M43', 'N57', 'R53', 'R62', 'T62', 'X42', 'X44')),]
cts <- as.matrix(subset(cts, select=-c(N56, X44, M43, N57, R53, R62, T62, X42, X44)))
coldata %>% dplyr::count(group)
all(colnames(cts) %in% rownames(coldata))

# Reorder cts/coldata by group for easy visualization in pcaexplorer
coldata <- coldata %>% arrange(group)
col.order <- rownames(coldata)
cts <- cts[,col.order]

# Remove gene-LOC designation from LOC #
rs <- (rownames(cts))
rs <- strsplit(rs, split = "LOC")
output <- matrix(ncol=1, nrow=length(rs))
# colnames(output) <- c("x","y","z")
for (i in 1:length(rs)){
  LOC_number <- as.data.frame(rs[[i]])
  add_val <- paste("LOC",LOC_number[3,1],sep="")
  output[i,] <- add_val
}
rownames(cts) <- output

```

# Create output folders
```{bash}
pwd
cd ..
pwd

# make output folder
mkdir output

# samples included _ variable analyzed (when 1 specified, compared to control)
mkdir output/all_treatments
mkdir output/control_ploidy
mkdir output/heat_ploidy
mkdir output/desiccation_ploidy
mkdir output/diploid_heat
mkdir output/diploid_desiccation
mkdir output/triploid_heat
mkdir output/triploid_desiccation
mkdir output/diploid
mkdir output/triploid

```
# ALL treatments, both ploidy
```{r ALL:TRT:ploidy, warning=FALSE, include=TRUE}

# Filter data
coldata_trt <- coldata # %>% filter(group == "A" | group == "B")
cts_trt     <- subset(cts, select=row.names(coldata_trt))

# Calculate DESeq object
dds <- DESeqDataSetFromMatrix(countData = cts_trt,
                              colData = coldata_trt,
                              design = ~ group)
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients

# Filtering: keep genes that have at least 10 counts across all samples - https://support.bioconductor.org/p/110307/
dds <- dds[rowSums(DESeq2::counts(dds)) >= 10,]

# Plot PCA
p <- pca(assay(vst(dds)), metadata = coldata_trt)
screeplot(p, axisLabSize = 18, titleLabSize = 22)
biplot(p, showLoadings = TRUE, colby = 'group',
    labSize = 5, pointSize = 5, sizeLoadingsNames = 5)

# A - diploid; control
# B - triploid; control
# C - diploid; heat
# D - triploid; heat
# E - diploid; desiccation
# F - triploid; desiccation

bp1 <- biplot(p, x = "PC1", y = "PC2",
              colby = 'group',
              colkey = c('A' = 'royalblue1' , 
                         'B' = 'green1'     ,
                         'C' = 'yellow2' ,
                         'D' = 'orange' ,
                         'E' = 'orangered1',
                         'F' = 'red3'),
              shape = 'ploidy',
              shapekey = c('diploid' = 19,  # circle
                           'triploid' = 17), # triangle
              pointSize = 4,
              showLoadings = FALSE,
              lab = NULL,
              drawConnectors = FALSE,
              ellipse = TRUE,
              ellipseLevel = 0.95,
              ellipseFill = FALSE,
              # ellipseFillKey = c('diploid' = 'blue', 'triploid' = 'green'),
              ellipseAlpha = 1/4,
              ellipseLineSize = 1,
              xlim = c(-150,150), ylim = c(-80, 50),
              gridlines.major = FALSE,
              gridlines.minor = FALSE,
              borderWidth = 1,
              legendPosition = 'top', legendLabSize = 16, legendIconSize = 6.0) + 
              theme(axis.text.x       = element_text(size=16,color="black"),
                    axis.text.y       = element_text(size=16,color="black"),
                    axis.ticks.x      = element_line(color="black"),
                    axis.ticks.y      = element_line(color="black"))
bp1

bp2 <- pairsplot(p,
                components = getComponents(p, c(1:5)),
                triangle = TRUE, trianglelabSize = 12,
                hline = 0, vline = 0,
                pointSize = 1,
                gridlines.major = FALSE, gridlines.minor = FALSE,
                colby = 'group',
                title = 'Pairs plot', plotaxes = FALSE,
                margingaps = unit(c(-0.01, -0.01, -0.01, -0.01), 'cm'))
bp2

# Plot pheatmap
# rld <- rlog(dds, blind=T)
# rld_mat <- assay(rld)
# pca <- prcomp(t(rld_mat))
# rld_cor <- cor(rld_mat)
# bp3 <- pheatmap(rld_cor, annotation = coldata_trt[,1:2])
# 
# bp3

setwd("..")
setwd("output/all_treatments")
ggexport(filename = "ALL-TREATMENTS-PCA.png",
         plot   = bp1,
         res    = 600,
         # device = "png",
         width  = 4000,
         height = 5000)

ggexport(filename = "ALL-TREATMENTS-PAIRS.png",
         plot   = bp2,
         res    = 600,
         width  = 6000,
         height = 6000)

ggexport(filename = "ALL-TREATMENTS-pheatmap.png",
         plot   = bp3,
         res    = 600,
         width  = 6000,
         height = 6000)

setwd("..")

# pcaExplorer(dds = dds_ploidy)


```

# ALL treatments, diploids only
```{r ALL:TRT:diploids, warning=FALSE, include=TRUE}

# Filter data
coldata_trt <- coldata %>% filter(group == "A" | group == "C" | group == "E")
cts_trt     <- subset(cts, select=row.names(coldata_trt))

# Calculate DESeq object
dds <- DESeqDataSetFromMatrix(countData = cts_trt,
                              colData = coldata_trt,
                              design = ~ group)
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients

# Filtering: keep genes that have at least 10 counts across 1/3 of the samples - https://support.bioconductor.org/p/110307/
keep <- rowSums(DESeq2::counts(dds) >= 10) >= ncol(cts_trt)/3
dds <- dds[keep,]

# Plot PCA
p <- pca(assay(vst(dds)), metadata = coldata_trt)
screeplot(p, axisLabSize = 18, titleLabSize = 22)
biplot(p, showLoadings = TRUE, colby = 'group',
    labSize = 5, pointSize = 5, sizeLoadingsNames = 5)

# A - diploid; control
# B - triploid; control
# C - diploid; heat
# D - triploid; heat
# E - diploid; desiccation
# F - triploid; desiccation

bp1 <- biplot(p, x = "PC1", y = "PC2",
              colby = 'group',
              colkey = c('A' = 'royalblue1' , 
                         'B' = 'green1'     ,
                         'C' = 'yellow2' ,
                         'D' = 'orange' ,
                         'E' = 'orangered1',
                         'F' = 'red3'),
              shape = 'ploidy',
              shapekey = c('diploid' = 19,  # circle
                           'triploid' = 17), # triangle
              pointSize = 4,
              showLoadings = FALSE,
              lab = NULL,
              drawConnectors = FALSE,
              ellipse = TRUE,
              ellipseLevel = 0.95,
              ellipseFill = FALSE,
              # ellipseFillKey = c('diploid' = 'blue', 'triploid' = 'green'),
              ellipseAlpha = 1/4,
              ellipseLineSize = 1,
              xlim = c(-150,140), ylim = c(-60, 40),
              gridlines.major = FALSE,
              gridlines.minor = FALSE,
              borderWidth = 1,
              legendPosition = 'top', legendLabSize = 16, legendIconSize = 6.0) + 
              theme(axis.text.x       = element_text(size=16,color="black"),
                    axis.text.y       = element_text(size=16,color="black"),
                    axis.ticks.x      = element_line(color="black"),
                    axis.ticks.y      = element_line(color="black"))
bp1

bp2 <- pairsplot(p,
                components = getComponents(p, c(1:5)),
                triangle = TRUE, trianglelabSize = 12,
                hline = 0, vline = 0,
                pointSize = 1,
                gridlines.major = FALSE, gridlines.minor = FALSE,
                colby = 'group',
                title = 'Pairs plot', plotaxes = FALSE,
                margingaps = unit(c(-0.01, -0.01, -0.01, -0.01), 'cm'))
bp2

# Plot pheatmap
# rld <- rlog(dds, blind=T)
# rld_mat <- assay(rld)
# pca <- prcomp(t(rld_mat))
# rld_cor <- cor(rld_mat)
# bp3 <- pheatmap(rld_cor, annotation = coldata_trt[,1:2])
# 
# bp3

setwd("..")
setwd("output/diploid")
ggexport(filename = "ALL-TREATMENTS-DIPLOID-PCA.png",
         plot   = bp1,
         res    = 600,
         # device = "png",
         width  = 4000,
         height = 5000)

ggexport(filename = "ALL-TREATMENTS-DIPLOID-PAIRS.png",
         plot   = bp2,
         res    = 600,
         width  = 6000,
         height = 6000)

# ggexport(filename = "ALL-TREATMENTS-DIPLOID-pheatmap.png",
#          plot   = bp3,
#          res    = 600,
#          width  = 6000,
#          height = 6000)

setwd("..")

# pcaExplorer(dds = dds_ploidy)


```


# ALL treatments, triploids only
```{r ALL:TRT:diploids, warning=FALSE, include=TRUE}

# Filter data
coldata_trt <- coldata %>% filter(group == "B" | group == "D" | group == "F")
cts_trt     <- subset(cts, select=row.names(coldata_trt))

# Calculate DESeq object
dds <- DESeqDataSetFromMatrix(countData = cts_trt,
                              colData = coldata_trt,
                              design = ~ group)
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients

# Filtering: keep genes that have at least 10 counts across all samples - https://support.bioconductor.org/p/110307/
dds <- dds[rowSums(DESeq2::counts(dds)) >= 10,]

# Plot PCA
p <- pca(assay(vst(dds)), metadata = coldata_trt)
screeplot(p, axisLabSize = 18, titleLabSize = 22)
biplot(p, showLoadings = TRUE, colby = 'group',
    labSize = 5, pointSize = 5, sizeLoadingsNames = 5)

# A - diploid; control
# B - triploid; control
# C - diploid; heat
# D - triploid; heat
# E - diploid; desiccation
# F - triploid; desiccation

bp1 <- biplot(p, x = "PC1", y = "PC2",
              colby = 'group',
              colkey = c('A' = 'royalblue1' , 
                         'B' = 'green1'     ,
                         'C' = 'yellow2' ,
                         'D' = 'orange' ,
                         'E' = 'orangered1',
                         'F' = 'red3'),
              shape = 'ploidy',
              shapekey = c('diploid' = 19,  # circle
                           'triploid' = 17), # triangle
              pointSize = 4,
              showLoadings = FALSE,
              lab = NULL,
              drawConnectors = FALSE,
              ellipse = TRUE,
              ellipseLevel = 0.95,
              ellipseFill = FALSE,
              # ellipseFillKey = c('diploid' = 'blue', 'triploid' = 'green'),
              ellipseAlpha = 1/4,
              ellipseLineSize = 1,
              xlim = c(-140,140), ylim = c(-60, 70),
              gridlines.major = FALSE,
              gridlines.minor = FALSE,
              borderWidth = 1,
              legendPosition = 'top', legendLabSize = 16, legendIconSize = 6.0) + 
              theme(axis.text.x       = element_text(size=16,color="black"),
                    axis.text.y       = element_text(size=16,color="black"),
                    axis.ticks.x      = element_line(color="black"),
                    axis.ticks.y      = element_line(color="black"))
bp1

bp2 <- pairsplot(p,
                components = getComponents(p, c(1:5)),
                triangle = TRUE, trianglelabSize = 12,
                hline = 0, vline = 0,
                pointSize = 1,
                gridlines.major = FALSE, gridlines.minor = FALSE,
                colby = 'group',
                title = 'Pairs plot', plotaxes = FALSE,
                margingaps = unit(c(-0.01, -0.01, -0.01, -0.01), 'cm'))
bp2

# Plot pheatmap
# rld <- rlog(dds, blind=T)
# rld_mat <- assay(rld)
# pca <- prcomp(t(rld_mat))
# rld_cor <- cor(rld_mat)
# bp3 <- pheatmap(rld_cor, annotation = coldata_trt[,1:2])
# 
# bp3

setwd("..")
setwd("output/triploid")
ggexport(filename = "ALL-TREATMENTS-TRIPLOID-PCA.png",
         plot   = bp1,
         res    = 600,
         # device = "png",
         width  = 4000,
         height = 5000)

ggexport(filename = "ALL-TREATMENTS-TRIPLOID-PAIRS.png",
         plot   = bp2,
         res    = 600,
         width  = 6000,
         height = 6000)

# ggexport(filename = "ALL-TREATMENTS-TRIPLOID-pheatmap.png",
#          plot   = bp3,
#          res    = 600,
#          width  = 6000,
#          height = 6000)

setwd("..")

# pcaExplorer(dds = dds_ploidy)


```



# CONTROLS only - effect of ploidy
```{r CONTROL:PLOIDY, warning=FALSE, include=TRUE}

# A - diploid; control
# B - triploid; control
# C - diploid; heat
# D - triploid; heat
# E - diploid; desiccation
# F - triploid; desiccation

# Filter data
coldata_trt <- coldata %>% filter(group == "A" | group == "B")
cts_trt     <- subset(cts, select=row.names(coldata_trt))

# Calculate DESeq object
dds <- DESeqDataSetFromMatrix(countData = cts_trt,
                              colData = coldata_trt,
                              design = ~ ploidy)
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients

# Filtering: keep genes that have at least 10 counts across all samples - https://support.bioconductor.org/p/110307/
dds <- dds[rowSums(DESeq2::counts(dds)) >= 10,]

# Generate Contrasts
contrast_list    <- c("ploidy", "triploid", "diploid")
res_table        <- results(dds, contrast=contrast_list, alpha = 0.05)
res_table_normal <- lfcShrink(dds,
                              coef=2, 
                              type="normal") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_apeglm <- lfcShrink(dds,
                              coef=2, 
                              type="apeglm") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_ashr   <- lfcShrink(dds,
                              coef=2, 
                              type="ashr")

# Plot MA
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)

pdf(file= "../output/control_ploidy/MA_plots.pdf" )
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
dev.off()

# Set cut.offs
padj.cutoff <- 0.05 # pvalues not produced with apeglm shrinkage estimator
svalue.cutoff <- 0.005 # 0.005 corresponds to pvalue of 0.05
lfc.cutoff <- 1.5 # fold-change cutoff, 1.5 or 2-fold is recommended

# Rearrange tables by gene
res_table <- res_table %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_normal <- res_table_normal %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_apeglm <- res_table_apeglm %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_ashr <- res_table_ashr %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

# How many genes are differentially expressed compared to Control?
colnames(LOCID_Roslin_gigas_gene_mito_table) <- c("gene","description","feature")
all_genes           <- left_join(res_table, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes           <- all_genes %>% arrange(-log2FoldChange)
all_genes$treatment <- "control_ploidy"
# all_genes_pvalue    <- cbind(gene=all_genes$gene,padj=all_genes$padj)

all_genes_normal           <- left_join(res_table_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_normal           <- all_genes_normal %>% arrange(-log2FoldChange)
all_genes_normal$treatment <- "control_ploidy"

all_genes_apeglm           <- left_join(res_table_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_apeglm           <- all_genes_apeglm %>% arrange(-log2FoldChange)
# all_genes_apeglm           <- left_join(all_genes_apeglm, as.data.frame(all_genes_pvalue), by = "gene")
all_genes_apeglm$treatment <- "control_ploidy"

all_genes_ashr           <- left_join(res_table_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_ashr           <- all_genes_ashr %>% arrange(-log2FoldChange)
all_genes_ashr$treatment <- "control_ploidy"

# Filter results by
sig_genes <- res_table %>%
  filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes <- left_join(sig_genes, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes <- sig_genes %>% arrange(-log2FoldChange)
sig_genes$treatment <- "control_ploidy" 

sig_genes_normal <- res_table_normal %>%
  filter(abs(log2FoldChange) > lfc.cutoff)
sig_genes_normal <- left_join(sig_genes_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_normal <- sig_genes_normal %>% arrange(-log2FoldChange)
sig_genes_normal$treatment <- "control_ploidy" 

sig_genes_apeglm <- res_table_apeglm %>%
  filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes_apeglm <- left_join(sig_genes_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_apeglm <- sig_genes_apeglm %>% arrange(-log2FoldChange)
sig_genes_apeglm$treatment <- "control_ploidy" 

sig_genes_ashr <- res_table_ashr %>%
  filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes_ashr <- left_join(sig_genes_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_ashr <- sig_genes_ashr %>% arrange(-log2FoldChange)
sig_genes_ashr$treatment <- "control_ploidy" 

# Save single gene expression counts to a data frame object, for each shrinkage estimator
single_gene_expression_unshrunk <- data.frame()
for (i in 1:nrow(sig_genes)){
  LOCID <- sig_genes$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_unshrunk <- rbind(single_gene_expression_unshrunk,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_normal <- data.frame()
for (i in 1:nrow(sig_genes_normal)){
  LOCID <- sig_genes_normal$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_normal <- rbind(single_gene_expression_normal,gene_counts_data)
  rm(gene_counts_data)
}


single_gene_expression_apeglm <- data.frame()
for (i in 1:nrow(sig_genes_apeglm)){
  LOCID <- sig_genes_apeglm$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_apeglm <- rbind(single_gene_expression_apeglm,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_ashr <- data.frame()
for (i in 1:nrow(sig_genes_ashr)){
  LOCID <- sig_genes_ashr$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_ashr <- rbind(single_gene_expression_ashr,gene_counts_data)
  rm(gene_counts_data)
}

write.table(single_gene_expression_unshrunk, file = "../output/control_ploidy/CONTROL_PLOIDY-unshrunk-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_normal, file = "../output/control_ploidy/CONTROL_PLOIDY-normal-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_apeglm, file = "../output/control_ploidy/CONTROL_PLOIDY-apeglm-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_ashr, file = "../output/control_ploidy/CONTROL_PLOIDY-ashr-single-gene-counts.csv", row.names = FALSE)

print(paste("# genes before filtering:", nrow(cts))) # count number of genes before filtering
print(paste("# genes after filtering:", length(dds))) # count number of genes after filtering
print(paste("# of genes dropped:", nrow(cts) - length(dds), sep=" ")) # count number of genes dropped
print(paste("# DEGs, all-genes:", nrow(all_genes)))
print(paste("# DEGs, unshrunken, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes)))
print(paste("# DEGs, normal shrinkage estimator, lfc = 1.5:", nrow(sig_genes_normal)))
print(paste("# DEGs, apeglm shrinkage estimator, svalue = 0.005, lfc = 1.5:", nrow(sig_genes_apeglm)))
print(paste("# DEGs, ashr shrinkage estimator, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes_ashr)))
gene_counts <- as.data.frame(c(nrow(cts),length(dds),(nrow(cts)-length(dds)),nrow(all_genes),
                            nrow(all_genes_normal),nrow(all_genes_apeglm),nrow(all_genes_ashr),
                            nrow(sig_genes),nrow(sig_genes_normal),
                            nrow(sig_genes_apeglm),nrow(sig_genes_ashr)))
row.names(gene_counts) <- c("genes_before_filtering","genes_after_filtering","genes_dropped",
                            "DEGs_all-genes","DEGs_all-genes-normal",
                            "DEGs_all-genes-apeglm","DEGs_all-genes-ashr",
                            "DEG_unshrunken-p0.05_lfc1.5","DEG_normal-lfc1.5",
                            "DEG_apeglm-s0.005_lfc1.5","DEG_ashr-p0.05_lfc1.5")
colnames(gene_counts) <- "count"

# Output tables of DEGs
write.table(all_genes,        file = "../output/control_ploidy/CONTROL_PLOIDY-ALL-DEG.csv",            row.names = FALSE)
write.table(all_genes_normal, file = "../output/control_ploidy/CONTROL_PLOIDY-ALL-DEG-normal.csv",     row.names = FALSE)
write.table(all_genes_apeglm, file = "../output/control_ploidy/CONTROL_PLOIDY-ALL-DEG-apeglm.csv",     row.names = FALSE)
write.table(all_genes_ashr,   file = "../output/control_ploidy/CONTROL_PLOIDY-ALL-DEG-ashr.csv",       row.names = FALSE)
write.table(sig_genes,        file = "../output/control_ploidy/CONTROL_PLOIDY-SIG-DEG-unshrunken.csv", row.names = FALSE)
write.table(sig_genes_normal, file = "../output/control_ploidy/CONTROL_PLOIDY-SIG-DEG-normal.csv",     row.names = FALSE)
write.table(sig_genes_apeglm, file = "../output/control_ploidy/CONTROL_PLOIDY-SIG-DEG-apeglm.csv",     row.names = FALSE)
write.table(sig_genes_ashr,   file = "../output/control_ploidy/CONTROL_PLOIDY-SIG-DEG-ashr.csv",       row.names = FALSE)
write.table(gene_counts,      file = "../output/control_ploidy/CONTROL_PLOIDY-gene-counts.csv",        row.names = TRUE)

# Plot PCA
# p <- pca(assay(vst(dds)), metadata = coldata_trt)
# screeplot(p, axisLabSize = 18, titleLabSize = 22)
# biplot(p, showLoadings = TRUE, colby = 'group',
#     labSize = 5, pointSize = 5, sizeLoadingsNames = 5)
# 
# bp1 <- biplot(p, x = "PC1", y = "PC2",
#               colby = 'ploidy',
#               colkey = c('diploid' = 'royalblue1', 'triploid' = 'green1'),
#               shape = 'ploidy',
#               shapekey = c('diploid' = 19, 'triploid' = 17),
#               pointSize = 4,
#               showLoadings = FALSE,
#               lab = NULL,
#               drawConnectors = FALSE,
#               ellipse = TRUE,
#               ellipseLevel = 0.95,
#               ellipseFill = FALSE,
#               # ellipseFillKey = c('diploid' = 'blue', 'triploid' = 'green'),
#               ellipseAlpha = 1/4,
#               ellipseLineSize = 1,
#               xlim = c(-125,145), ylim = c(-100, 80),
#               gridlines.major = FALSE,
#               gridlines.minor = FALSE,
#               borderWidth = 1,
#               legendPosition = 'top', legendLabSize = 16, legendIconSize = 6.0) + 
#               theme(axis.text.x       = element_text(size=16,color="black"),
#                     axis.text.y       = element_text(size=16,color="black"),
#                     axis.ticks.x      = element_line(color="black"),
#                     axis.ticks.y      = element_line(color="black"))
# bp1
# 
# bp2 <- pairsplot(p,
#                 components = getComponents(p, c(1:5)),
#                 triangle = TRUE, trianglelabSize = 12,
#                 hline = 0, vline = 0,
#                 pointSize = 1,
#                 gridlines.major = FALSE, gridlines.minor = FALSE,
#                 colby = 'ploidy',
#                 title = 'Pairs plot', plotaxes = FALSE,
#                 margingaps = unit(c(-0.01, -0.01, -0.01, -0.01), 'cm'))
# bp2
# 
# # Plot pheatmap
# 
# rld <- rlog(dds, blind=T)
# rld_mat <- assay(rld)
# pca <- prcomp(t(rld_mat))
# rld_cor <- cor(rld_mat)
# bp3 <- pheatmap(rld_cor, annotation = coldata_trt[,1:2])
# 
# bp3
# 
# setwd("..")
# setwd("output/control_ploidy")
# ggexport(filename = "CONTROL-PLOIDY-PCA.png",
#          plot   = bp1,
#          res    = 600,
#          # device = "png",
#          width  = 4000,
#          height = 5000)
# 
# ggexport(filename = "CONTROL-PLOIDY-PAIRS.png",
#          plot   = bp2,
#          res    = 600,
#          width  = 6000,
#          height = 6000)
# 
# ggexport(filename = "CONTROL-PLOIDY-pheatmap.png",
#          plot   = bp3,
#          res    = 600,
#          width  = 6000,
#          height = 6000)
# 
# setwd("..")

# pcaExplorer(dds = dds_ploidy)


```

# HEAT only - effect of ploidy
```{r HEAT:PLOIDY, warning=FALSE, include=TRUE}

# A - diploid; control
# B - triploid; control
# C - diploid; heat
# D - triploid; heat
# E - diploid; desiccation
# F - triploid; desiccation

# Filter data
coldata_trt <- coldata %>% filter(group == "C" | group == "D")
cts_trt     <- subset(cts, select=row.names(coldata_trt))

# Calculate DESeq object
dds <- DESeqDataSetFromMatrix(countData = cts_trt,
                              colData = coldata_trt,
                              design = ~ ploidy)
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients

# Filtering: keep genes that have at least 10 counts across all samples - https://support.bioconductor.org/p/110307/
dds <- dds[rowSums(DESeq2::counts(dds)) >= 10,]

# Generate Contrasts
contrast_list    <- c("ploidy", "triploid", "diploid")
res_table        <- results(dds, contrast=contrast_list, alpha = 0.05)
res_table_normal <- lfcShrink(dds,
                              coef=2, 
                              type="normal") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_apeglm <- lfcShrink(dds,
                              coef=2, 
                              type="apeglm") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_ashr   <- lfcShrink(dds,
                              coef=2, 
                              type="ashr")

# Plot MA
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)

pdf(file= "../output/heat_ploidy/MA_plots.pdf" )
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
dev.off()

# Set cut.offs
padj.cutoff <- 0.05 # pvalues not produced with apeglm shrinkage estimator
svalue.cutoff <- 0.005 # 0.005 corresponds to pvalue of 0.05
lfc.cutoff <- 1.5 # fold-change cutoff, 1.5 or 2-fold is recommended

# Rearrange tables by gene
res_table <- res_table %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_normal <- res_table_normal %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_apeglm <- res_table_apeglm %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_ashr <- res_table_ashr %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

# How many genes are differentially expressed compared to Control?
colnames(LOCID_Roslin_gigas_gene_mito_table) <- c("gene","description","feature")
all_genes           <- left_join(res_table, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes           <- all_genes %>% arrange(-log2FoldChange)
all_genes$treatment <- "heat_ploidy"
# all_genes_pvalue    <- cbind(gene=all_genes$gene,padj=all_genes$padj)

all_genes_normal           <- left_join(res_table_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_normal           <- all_genes_normal %>% arrange(-log2FoldChange)
all_genes_normal$treatment <- "heat_ploidy"

all_genes_apeglm           <- left_join(res_table_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_apeglm           <- all_genes_apeglm %>% arrange(-log2FoldChange)
# all_genes_apeglm           <- left_join(all_genes_apeglm, as.data.frame(all_genes_pvalue), by = "gene")
all_genes_apeglm$treatment <- "heat_ploidy"

all_genes_ashr           <- left_join(res_table_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_ashr           <- all_genes_ashr %>% arrange(-log2FoldChange)
all_genes_ashr$treatment <- "heat_ploidy"

# Filter results by
sig_genes <- res_table %>%
  filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes <- left_join(sig_genes, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes <- sig_genes %>% arrange(-log2FoldChange)
sig_genes$treatment <- "heat_ploidy" 

sig_genes_normal <- res_table_normal %>%
  filter(abs(log2FoldChange) > lfc.cutoff)
sig_genes_normal <- left_join(sig_genes_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_normal <- sig_genes_normal %>% arrange(-log2FoldChange)
sig_genes_normal$treatment <- "heat_ploidy" 

sig_genes_apeglm <- res_table_apeglm %>%
  filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes_apeglm <- left_join(sig_genes_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_apeglm <- sig_genes_apeglm %>% arrange(-log2FoldChange)
sig_genes_apeglm$treatment <- "heat_ploidy" 

sig_genes_ashr <- res_table_ashr %>%
  filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes_ashr <- left_join(sig_genes_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_ashr <- sig_genes_ashr %>% arrange(-log2FoldChange)
sig_genes_ashr$treatment <- "heat_ploidy" 

# Save single gene expression counts to a data frame object, for each shrinkage estimator
single_gene_expression_unshrunk <- data.frame()
for (i in 1:nrow(sig_genes)){
  LOCID <- sig_genes$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_unshrunk <- rbind(single_gene_expression_unshrunk,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_normal <- data.frame()
for (i in 1:nrow(sig_genes_normal)){
  LOCID <- sig_genes_normal$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_normal <- rbind(single_gene_expression_normal,gene_counts_data)
  rm(gene_counts_data)
}


single_gene_expression_apeglm <- data.frame()
for (i in 1:nrow(sig_genes_apeglm)){
  LOCID <- sig_genes_apeglm$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_apeglm <- rbind(single_gene_expression_apeglm,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_ashr <- data.frame()
for (i in 1:nrow(sig_genes_ashr)){
  LOCID <- sig_genes_ashr$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_ashr <- rbind(single_gene_expression_ashr,gene_counts_data)
  rm(gene_counts_data)
}

write.table(single_gene_expression_unshrunk, file = "../output/heat_ploidy/HEAT_PLOIDY-unshrunk-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_normal, file = "../output/heat_ploidy/HEAT_PLOIDY-normal-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_apeglm, file = "../output/heat_ploidy/HEAT_PLOIDY-apeglm-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_ashr, file = "../output/heat_ploidy/HEAT_PLOIDY-ashr-single-gene-counts.csv", row.names = FALSE)

print(paste("# genes before filtering:", nrow(cts))) # count number of genes before filtering
print(paste("# genes after filtering:", length(dds))) # count number of genes after filtering
print(paste("# of genes dropped:", nrow(cts) - length(dds), sep=" ")) # count number of genes dropped
print(paste("# DEGs, all-genes:", nrow(all_genes)))
print(paste("# DEGs, unshrunken, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes)))
print(paste("# DEGs, normal shrinkage estimator, lfc = 1.5:", nrow(sig_genes_normal)))
print(paste("# DEGs, apeglm shrinkage estimator, svalue = 0.005, lfc = 1.5:", nrow(sig_genes_apeglm)))
print(paste("# DEGs, ashr shrinkage estimator, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes_ashr)))
gene_counts <- as.data.frame(c(nrow(cts),length(dds),(nrow(cts)-length(dds)),nrow(all_genes),
                            nrow(all_genes_normal),nrow(all_genes_apeglm),nrow(all_genes_ashr),
                            nrow(sig_genes),nrow(sig_genes_normal),
                            nrow(sig_genes_apeglm),nrow(sig_genes_ashr)))
row.names(gene_counts) <- c("genes_before_filtering","genes_after_filtering","genes_dropped",
                            "DEGs_all-genes","DEGs_all-genes-normal",
                            "DEGs_all-genes-apeglm","DEGs_all-genes-ashr",
                            "DEG_unshrunken-p0.05_lfc1.5","DEG_normal-lfc1.5",
                            "DEG_apeglm-s0.005_lfc1.5","DEG_ashr-p0.05_lfc1.5")
colnames(gene_counts) <- "count"

# Output tables of DEGs
write.table(all_genes,        file = "../output/heat_ploidy/HEAT_PLOIDY-ALL-DEG.csv",            row.names = FALSE)
write.table(all_genes_normal, file = "../output/heat_ploidy/HEAT_PLOIDY-ALL-DEG-normal.csv",     row.names = FALSE)
write.table(all_genes_apeglm, file = "../output/heat_ploidy/HEAT_PLOIDY-ALL-DEG-apeglm.csv",     row.names = FALSE)
write.table(all_genes_ashr,   file = "../output/heat_ploidy/HEAT_PLOIDY-ALL-DEG-ashr.csv",       row.names = FALSE)
write.table(sig_genes,        file = "../output/heat_ploidy/HEAT_PLOIDY-SIG-DEG-unshrunken.csv", row.names = FALSE)
write.table(sig_genes_normal, file = "../output/heat_ploidy/HEAT_PLOIDY-SIG-DEG-normal.csv",     row.names = FALSE)
write.table(sig_genes_apeglm, file = "../output/heat_ploidy/HEAT_PLOIDY-SIG-DEG-apeglm.csv",     row.names = FALSE)
write.table(sig_genes_ashr,   file = "../output/heat_ploidy/HEAT_PLOIDY-SIG-DEG-ashr.csv",       row.names = FALSE)
write.table(gene_counts,      file = "../output/heat_ploidy/HEAT_PLOIDY-gene-counts.csv",        row.names = TRUE)

# Plot PCA
# p <- pca(assay(vst(dds)), metadata = coldata_trt)
# screeplot(p, axisLabSize = 18, titleLabSize = 22)
# biplot(p, showLoadings = TRUE, colby = 'group',
#     labSize = 5, pointSize = 5, sizeLoadingsNames = 5)
# 
# bp1 <- biplot(p, x = "PC1", y = "PC2",
#               colby = 'group',
#               colkey = c('A' = 'royalblue1' , 
#                          'B' = 'green1'     ,
#                          'C' = 'yellow2' ,
#                          'D' = 'orange' ,
#                          'E' = 'orangered1',
#                          'F' = 'red3'),
#               shape = 'ploidy',
#               shapekey = c('diploid' = 19, 'triploid' = 17),
#               pointSize = 4,
#               showLoadings = FALSE,
#               lab = NULL,
#               drawConnectors = FALSE,
#               ellipse = TRUE,
#               ellipseLevel = 0.95,
#               ellipseFill = FALSE,
#               # ellipseFillKey = c('diploid' = 'blue', 'triploid' = 'green'),
#               ellipseAlpha = 1/4,
#               ellipseLineSize = 1,
#               xlim = c(-100,110), ylim = c(-50, 60),
#               gridlines.major = FALSE,
#               gridlines.minor = FALSE,
#               borderWidth = 1,
#               legendPosition = 'top', legendLabSize = 16, legendIconSize = 6.0) + 
#               theme(axis.text.x       = element_text(size=16,color="black"),
#                     axis.text.y       = element_text(size=16,color="black"),
#                     axis.ticks.x      = element_line(color="black"),
#                     axis.ticks.y      = element_line(color="black"))
# bp1
# 
# bp2 <- pairsplot(p,
#                 components = getComponents(p, c(1:5)),
#                 triangle = TRUE, trianglelabSize = 12,
#                 hline = 0, vline = 0,
#                 pointSize = 1,
#                 gridlines.major = FALSE, gridlines.minor = FALSE,
#                 colby = 'ploidy',
#                 title = 'Pairs plot', plotaxes = FALSE,
#                 margingaps = unit(c(-0.01, -0.01, -0.01, -0.01), 'cm'))
# bp2
# 
# 
# # Plot pheatmap
# 
# rld <- rlog(dds, blind=T)
# rld_mat <- assay(rld)
# pca <- prcomp(t(rld_mat))
# rld_cor <- cor(rld_mat)
# bp3 <- pheatmap(rld_cor, annotation = coldata_trt[,1:2])
# 
# bp3
# 
# 
# 
# 
# setwd("..")
# setwd("output/heat_ploidy")
# ggexport(filename = "HEAT-PLOIDY-PCA.png",
#          plot   = bp1,
#          res    = 600,
#          # device = "png",
#          width  = 4000,
#          height = 5000)
# 
# ggexport(filename = "HEAT-PLOIDY-PAIRS.png",
#          plot   = bp2,
#          res    = 600,
#          width  = 6000,
#          height = 6000)
# 
# ggexport(filename = "HEAT-PLOIDY-pheatmap.png",
#          plot   = bp3,
#          res    = 600,
#          width  = 6000,
#          height = 6000)
# setwd("..")

# pcaExplorer(dds = dds_ploidy)


```

# DESICCATION only - effect of ploidy
```{r DESICCATION:PLOIDY, warning=FALSE, include=TRUE}

# A - diploid; control
# B - triploid; control
# C - diploid; heat
# D - triploid; heat
# E - diploid; desiccation
# F - triploid; desiccation

# Filter data
coldata_trt <- coldata %>% filter(group == "E" | group == "F")
cts_trt     <- subset(cts, select=row.names(coldata_trt))

# Calculate DESeq object
dds <- DESeqDataSetFromMatrix(countData = cts_trt,
                              colData = coldata_trt,
                              design = ~ ploidy)
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients

# Filtering: keep genes that have at least 10 counts across all samples - https://support.bioconductor.org/p/110307/
dds <- dds[rowSums(DESeq2::counts(dds)) >= 10,]

# Generate Contrasts
contrast_list    <- c("ploidy", "triploid", "diploid")
res_table        <- results(dds, contrast=contrast_list, alpha = 0.05)
res_table_normal <- lfcShrink(dds,
                              coef=2, 
                              type="normal") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_apeglm <- lfcShrink(dds,
                              coef=2, 
                              type="apeglm") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_ashr   <- lfcShrink(dds,
                              coef=2, 
                              type="ashr")

# Plot MA
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)

pdf(file= "../output/desiccation_ploidy/MA_plots.pdf" )
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
dev.off()

# Set cut.offs
padj.cutoff <- 0.05 # pvalues not produced with apeglm shrinkage estimator
svalue.cutoff <- 0.005 # 0.005 corresponds to pvalue of 0.05
lfc.cutoff <- 1.5 # fold-change cutoff, 1.5 or 2-fold is recommended

# Rearrange tables by gene
res_table <- res_table %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_normal <- res_table_normal %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_apeglm <- res_table_apeglm %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_ashr <- res_table_ashr %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

# How many genes are differentially expressed compared to Control?
colnames(LOCID_Roslin_gigas_gene_mito_table) <- c("gene","description","feature")
all_genes           <- left_join(res_table, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes           <- all_genes %>% arrange(-log2FoldChange)
all_genes$treatment <- "desiccation_ploidy"
# all_genes_pvalue    <- cbind(gene=all_genes$gene,padj=all_genes$padj)

all_genes_normal           <- left_join(res_table_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_normal           <- all_genes_normal %>% arrange(-log2FoldChange)
all_genes_normal$treatment <- "desiccation_ploidy"

all_genes_apeglm           <- left_join(res_table_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_apeglm           <- all_genes_apeglm %>% arrange(-log2FoldChange)
# all_genes_apeglm           <- left_join(all_genes_apeglm, as.data.frame(all_genes_pvalue), by = "gene")
all_genes_apeglm$treatment <- "desiccation_ploidy"

all_genes_ashr           <- left_join(res_table_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_ashr           <- all_genes_ashr %>% arrange(-log2FoldChange)
all_genes_ashr$treatment <- "desiccation_ploidy"

# Filter results by
sig_genes <- res_table %>%
  filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes <- left_join(sig_genes, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes <- sig_genes %>% arrange(-log2FoldChange)
sig_genes$treatment <- "desiccation_ploidy" 

sig_genes_normal <- res_table_normal %>%
  filter(abs(log2FoldChange) > lfc.cutoff)
sig_genes_normal <- left_join(sig_genes_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_normal <- sig_genes_normal %>% arrange(-log2FoldChange)
sig_genes_normal$treatment <- "desiccation_ploidy" 

sig_genes_apeglm <- res_table_apeglm %>%
  filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes_apeglm <- left_join(sig_genes_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_apeglm <- sig_genes_apeglm %>% arrange(-log2FoldChange)
sig_genes_apeglm$treatment <- "desiccation_ploidy" 

sig_genes_ashr <- res_table_ashr %>%
  filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes_ashr <- left_join(sig_genes_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_ashr <- sig_genes_ashr %>% arrange(-log2FoldChange)
sig_genes_ashr$treatment <- "desiccation_ploidy" 

# Save single gene expression counts to a data frame object, for each shrinkage estimator
single_gene_expression_unshrunk <- data.frame()
for (i in 1:nrow(sig_genes)){
  LOCID <- sig_genes$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_unshrunk <- rbind(single_gene_expression_unshrunk,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_normal <- data.frame()
for (i in 1:nrow(sig_genes_normal)){
  LOCID <- sig_genes_normal$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_normal <- rbind(single_gene_expression_normal,gene_counts_data)
  rm(gene_counts_data)
}


single_gene_expression_apeglm <- data.frame()
for (i in 1:nrow(sig_genes_apeglm)){
  LOCID <- sig_genes_apeglm$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_apeglm <- rbind(single_gene_expression_apeglm,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_ashr <- data.frame()
for (i in 1:nrow(sig_genes_ashr)){
  LOCID <- sig_genes_ashr$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_ashr <- rbind(single_gene_expression_ashr,gene_counts_data)
  rm(gene_counts_data)
}

write.table(single_gene_expression_unshrunk, file = "../output/desiccation_ploidy/DESICCATION_PLOIDY-unshrunk-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_normal, file = "../output/desiccation_ploidy/DESICCATION_PLOIDY-normal-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_apeglm, file = "../output/desiccation_ploidy/DESICCATION_PLOIDY-apeglm-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_ashr, file = "../output/desiccation_ploidy/DESICCATION_PLOIDY-ashr-single-gene-counts.csv", row.names = FALSE)

print(paste("# genes before filtering:", nrow(cts))) # count number of genes before filtering
print(paste("# genes after filtering:", length(dds))) # count number of genes after filtering
print(paste("# of genes dropped:", nrow(cts) - length(dds), sep=" ")) # count number of genes dropped
print(paste("# DEGs, all-genes:", nrow(all_genes)))
print(paste("# DEGs, unshrunken, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes)))
print(paste("# DEGs, normal shrinkage estimator, lfc = 1.5:", nrow(sig_genes_normal)))
print(paste("# DEGs, apeglm shrinkage estimator, svalue = 0.005, lfc = 1.5:", nrow(sig_genes_apeglm)))
print(paste("# DEGs, ashr shrinkage estimator, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes_ashr)))
gene_counts <- as.data.frame(c(nrow(cts),length(dds),(nrow(cts)-length(dds)),nrow(all_genes),
                            nrow(all_genes_normal),nrow(all_genes_apeglm),nrow(all_genes_ashr),
                            nrow(sig_genes),nrow(sig_genes_normal),
                            nrow(sig_genes_apeglm),nrow(sig_genes_ashr)))
row.names(gene_counts) <- c("genes_before_filtering","genes_after_filtering","genes_dropped",
                            "DEGs_all-genes","DEGs_all-genes-normal",
                            "DEGs_all-genes-apeglm","DEGs_all-genes-ashr",
                            "DEG_unshrunken-p0.05_lfc1.5","DEG_normal-lfc1.5",
                            "DEG_apeglm-s0.005_lfc1.5","DEG_ashr-p0.05_lfc1.5")
colnames(gene_counts) <- "count"

# Output tables of DEGs
write.table(all_genes,        file = "../output/desiccation_ploidy/DESICCATION_PLOIDY-ALL-DEG.csv",            row.names = FALSE)
write.table(all_genes_normal, file = "../output/desiccation_ploidy/DESICCATION_PLOIDY-ALL-DEG-normal.csv",     row.names = FALSE)
write.table(all_genes_apeglm, file = "../output/desiccation_ploidy/DESICCATION_PLOIDY-ALL-DEG-apeglm.csv",     row.names = FALSE)
write.table(all_genes_ashr,   file = "../output/desiccation_ploidy/DESICCATION_PLOIDY-ALL-DEG-ashr.csv",       row.names = FALSE)
write.table(sig_genes,        file = "../output/desiccation_ploidy/DESICCATION_PLOIDY-SIG-DEG-unshrunken.csv", row.names = FALSE)
write.table(sig_genes_normal, file = "../output/desiccation_ploidy/DESICCATION_PLOIDY-SIG-DEG-normal.csv",     row.names = FALSE)
write.table(sig_genes_apeglm, file = "../output/desiccation_ploidy/DESICCATION_PLOIDY-SIG-DEG-apeglm.csv",     row.names = FALSE)
write.table(sig_genes_ashr,   file = "../output/desiccation_ploidy/DESICCATION_PLOIDY-SIG-DEG-ashr.csv",       row.names = FALSE)
write.table(gene_counts,      file = "../output/desiccation_ploidy/DESICCATION_PLOIDY-gene-counts.csv",        row.names = TRUE)


# # Plot PCA
# p <- pca(assay(vst(dds)), metadata = coldata_trt)
# screeplot(p, axisLabSize = 18, titleLabSize = 22)
# biplot(p, showLoadings = TRUE, colby = 'group',
#     labSize = 5, pointSize = 5, sizeLoadingsNames = 5)
# 
# bp1 <- biplot(p, x = "PC1", y = "PC2",
#               colby = 'group',
#               colkey = c('A' = 'royalblue1' , 
#                          'B' = 'green1'     ,
#                          'C' = 'yellow2' ,
#                          'D' = 'orange' ,
#                          'E' = 'orangered1',
#                          'F' = 'red3'),
#               shape = 'ploidy',
#               shapekey = c('diploid' = 19, 'triploid' = 17),
#               pointSize = 4,
#               showLoadings = FALSE,
#               lab = NULL,
#               drawConnectors = FALSE,
#               ellipse = TRUE,
#               ellipseLevel = 0.95,
#               ellipseFill = FALSE,
#               # ellipseFillKey = c('diploid' = 'blue', 'triploid' = 'green'),
#               ellipseAlpha = 1/4,
#               ellipseLineSize = 1,
#               xlim = c(-130,150), ylim = c(-40, 60),
#               gridlines.major = FALSE,
#               gridlines.minor = FALSE,
#               borderWidth = 1,
#               legendPosition = 'top', legendLabSize = 16, legendIconSize = 6.0) + 
#               theme(axis.text.x       = element_text(size=16,color="black"),
#                     axis.text.y       = element_text(size=16,color="black"),
#                     axis.ticks.x      = element_line(color="black"),
#                     axis.ticks.y      = element_line(color="black"))
# bp1
# 
# bp2 <- pairsplot(p,
#                 components = getComponents(p, c(1:5)),
#                 triangle = TRUE, trianglelabSize = 12,
#                 hline = 0, vline = 0,
#                 pointSize = 1,
#                 gridlines.major = FALSE, gridlines.minor = FALSE,
#                 colby = 'ploidy',
#                 title = 'Pairs plot', plotaxes = FALSE,
#                 margingaps = unit(c(-0.01, -0.01, -0.01, -0.01), 'cm'))
# bp2
# 
# # Plot pheatmap
# # rld <- rlog(dds, blind=T)
# # rld_mat <- assay(rld)
# # pca <- prcomp(t(rld_mat))
# # rld_cor <- cor(rld_mat)
# # bp3 <- pheatmap(rld_cor, annotation = coldata_trt[,1:2])
# # 
# # bp3
# 
# 
# 
# setwd("..")
# setwd("output/desiccation_ploidy")
# ggexport(filename = "DESICCATION-PLOIDY-PCA.png",
#          plot   = bp1,
#          res    = 600,
#          # device = "png",
#          width  = 4000,
#          height = 5000)
# 
# ggexport(filename = "DESICCATION-PLOIDY-PAIRS.png",
#          plot   = bp2,
#          res    = 600,
#          width  = 6000,
#          height = 6000)
# 
# # ggexport(filename = "DESICCATION-PLOIDY-pheatmap.png",
# #          plot   = bp3,
# #          res    = 600,
# #          width  = 6000,
# #          height = 6000)
# 
# setwd("..")

# pcaExplorer(dds = dds_ploidy)


```

# DIPLOID only - effect of temperature
```{r DIPLOID:HEAT, warning=FALSE, include=TRUE}

# A - diploid; control
# B - triploid; control
# C - diploid; heat
# D - triploid; heat
# E - diploid; desiccation
# F - triploid; desiccation

# Filter data
coldata_trt <- coldata %>% filter(group == "A" | group == "C")
cts_trt     <- subset(cts, select=row.names(coldata_trt))

# Calculate DESeq object
dds <- DESeqDataSetFromMatrix(countData = cts_trt,
                              colData = coldata_trt,
                              design = ~ condition)
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients

# Filtering: keep genes that have at least 10 counts across all samples - https://support.bioconductor.org/p/110307/
dds <- dds[rowSums(DESeq2::counts(dds)) >= 10,]

# Generate Contrasts
contrast_list    <- c("condition", "heat", "control") # factor, treatment group, control
res_table        <- results(dds, contrast=contrast_list, alpha = 0.05)
res_table_normal <- lfcShrink(dds,
                              coef=2, 
                              type="normal") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_apeglm <- lfcShrink(dds,
                              coef=2, 
                              type="apeglm") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_ashr   <- lfcShrink(dds,
                              coef=2, 
                              type="ashr")

# Plot MA
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)

pdf(file= "../output/diploid_heat/MA_plots.pdf" )
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
dev.off()

# Set cut.offs
padj.cutoff <- 0.05 # pvalues not produced with apeglm shrinkage estimator
svalue.cutoff <- 0.005 # 0.005 corresponds to pvalue of 0.05
lfc.cutoff <- 1.5 # fold-change cutoff, 1.5 or 2-fold is recommended

# Rearrange tables by gene
res_table <- res_table %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_normal <- res_table_normal %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_apeglm <- res_table_apeglm %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_ashr <- res_table_ashr %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

# How many genes are differentially expressed compared to Control?
colnames(LOCID_Roslin_gigas_gene_mito_table) <- c("gene","description","feature")
all_genes           <- left_join(res_table, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes           <- all_genes %>% arrange(-log2FoldChange)
all_genes$treatment <- "diploid_heat"
# all_genes_pvalue    <- cbind(gene=all_genes$gene,padj=all_genes$padj)

all_genes_normal           <- left_join(res_table_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_normal           <- all_genes_normal %>% arrange(-log2FoldChange)
all_genes_normal$treatment <- "diploid_heat"

all_genes_apeglm           <- left_join(res_table_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_apeglm           <- all_genes_apeglm %>% arrange(-log2FoldChange)
# all_genes_apeglm           <- left_join(all_genes_apeglm, as.data.frame(all_genes_pvalue), by = "gene")
all_genes_apeglm$treatment <- "diploid_heat"

all_genes_ashr           <- left_join(res_table_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_ashr           <- all_genes_ashr %>% arrange(-log2FoldChange)
all_genes_ashr$treatment <- "diploid_heat"

# Filter results by
sig_genes <- res_table %>%
  filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes <- left_join(sig_genes, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes <- sig_genes %>% arrange(-log2FoldChange)
sig_genes$treatment <- "diploid_heat" 

sig_genes_normal <- res_table_normal %>%
  filter(abs(log2FoldChange) > lfc.cutoff)
sig_genes_normal <- left_join(sig_genes_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_normal <- sig_genes_normal %>% arrange(-log2FoldChange)
sig_genes_normal$treatment <- "diploid_heat" 

sig_genes_apeglm <- res_table_apeglm %>%
  filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes_apeglm <- left_join(sig_genes_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_apeglm <- sig_genes_apeglm %>% arrange(-log2FoldChange)
sig_genes_apeglm$treatment <- "diploid_heat" 

sig_genes_ashr <- res_table_ashr %>%
  filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes_ashr <- left_join(sig_genes_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_ashr <- sig_genes_ashr %>% arrange(-log2FoldChange)
sig_genes_ashr$treatment <- "diploid_heat" 

# Save single gene expression counts to a data frame object, for each shrinkage estimator
single_gene_expression_unshrunk <- data.frame()
for (i in 1:nrow(sig_genes)){
  LOCID <- sig_genes$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_unshrunk <- rbind(single_gene_expression_unshrunk,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_normal <- data.frame()
for (i in 1:nrow(sig_genes_normal)){
  LOCID <- sig_genes_normal$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_normal <- rbind(single_gene_expression_normal,gene_counts_data)
  rm(gene_counts_data)
}


single_gene_expression_apeglm <- data.frame()
for (i in 1:nrow(sig_genes_apeglm)){
  LOCID <- sig_genes_apeglm$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_apeglm <- rbind(single_gene_expression_apeglm,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_ashr <- data.frame()
for (i in 1:nrow(sig_genes_ashr)){
  LOCID <- sig_genes_ashr$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_ashr <- rbind(single_gene_expression_ashr,gene_counts_data)
  rm(gene_counts_data)
}

write.table(single_gene_expression_unshrunk, file = "../output/diploid_heat/DIPLOID_HEAT-unshrunk-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_normal, file = "../output/diploid_heat/DIPLOID_HEAT-normal-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_apeglm, file = "../output/diploid_heat/DIPLOID_HEAT-apeglm-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_ashr, file = "../output/diploid_heat/DIPLOID_HEAT-ashr-single-gene-counts.csv", row.names = FALSE)

print(paste("# genes before filtering:", nrow(cts))) # count number of genes before filtering
print(paste("# genes after filtering:", length(dds))) # count number of genes after filtering
print(paste("# of genes dropped:", nrow(cts) - length(dds), sep=" ")) # count number of genes dropped
print(paste("# DEGs, all-genes:", nrow(all_genes)))
print(paste("# DEGs, unshrunken, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes)))
print(paste("# DEGs, normal shrinkage estimator, lfc = 1.5:", nrow(sig_genes_normal)))
print(paste("# DEGs, apeglm shrinkage estimator, svalue = 0.005, lfc = 1.5:", nrow(sig_genes_apeglm)))
print(paste("# DEGs, ashr shrinkage estimator, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes_ashr)))
gene_counts <- as.data.frame(c(nrow(cts),length(dds),(nrow(cts)-length(dds)),nrow(all_genes),
                            nrow(all_genes_normal),nrow(all_genes_apeglm),nrow(all_genes_ashr),
                            nrow(sig_genes),nrow(sig_genes_normal),
                            nrow(sig_genes_apeglm),nrow(sig_genes_ashr)))
row.names(gene_counts) <- c("genes_before_filtering","genes_after_filtering","genes_dropped",
                            "DEGs_all-genes","DEGs_all-genes-normal",
                            "DEGs_all-genes-apeglm","DEGs_all-genes-ashr",
                            "DEG_unshrunken-p0.05_lfc1.5","DEG_normal-lfc1.5",
                            "DEG_apeglm-s0.005_lfc1.5","DEG_ashr-p0.05_lfc1.5")
colnames(gene_counts) <- "count"

# Output tables of DEGs
write.table(all_genes,        file = "../output/diploid_heat/DIPLOID_HEAT-ALL-DEG.csv",            row.names = FALSE)
write.table(all_genes_normal, file = "../output/diploid_heat/DIPLOID_HEAT-ALL-DEG-normal.csv",     row.names = FALSE)
write.table(all_genes_apeglm, file = "../output/diploid_heat/DIPLOID_HEAT-ALL-DEG-apeglm.csv",     row.names = FALSE)
write.table(all_genes_ashr,   file = "../output/diploid_heat/DIPLOID_HEAT-ALL-DEG-ashr.csv",       row.names = FALSE)
write.table(sig_genes,        file = "../output/diploid_heat/DIPLOID_HEAT-SIG-DEG-unshrunken.csv", row.names = FALSE)
write.table(sig_genes_normal, file = "../output/diploid_heat/DIPLOID_HEAT-SIG-DEG-normal.csv",     row.names = FALSE)
write.table(sig_genes_apeglm, file = "../output/diploid_heat/DIPLOID_HEAT-SIG-DEG-apeglm.csv",     row.names = FALSE)
write.table(sig_genes_ashr,   file = "../output/diploid_heat/DIPLOID_HEAT-SIG-DEG-ashr.csv",       row.names = FALSE)
write.table(gene_counts,      file = "../output/diploid_heat/DIPLOID_HEAT-gene-counts.csv",        row.names = TRUE)

# # Plot PCA
# p <- pca(assay(vst(dds)), metadata = coldata_trt)
# screeplot(p, axisLabSize = 18, titleLabSize = 22)
# biplot(p, showLoadings = TRUE, colby = 'group',
#     labSize = 5, pointSize = 5, sizeLoadingsNames = 5)
# 
# bp1 <- biplot(p, x = "PC1", y = "PC2",
#               colby = 'group',
#               colkey = c('A' = 'royalblue1' , 
#                          'B' = 'green1'     ,
#                          'C' = 'yellow2' ,
#                          'D' = 'orange' ,
#                          'E' = 'orangered1',
#                          'F' = 'red3'),
#               shape = 'condition',
#               shapekey = c('control' = 19, 'heat' = 19),
#               pointSize = 4,
#               showLoadings = FALSE,
#               lab = NULL,
#               drawConnectors = FALSE,
#               ellipse = TRUE,
#               ellipseLevel = 0.95,
#               ellipseFill = FALSE,
#               ellipseFillKey = c('control' = 'royalblue1', 'heat' = 'orangered1'),
#               ellipseAlpha = 1/4,
#               ellipseLineSize = 1,
#               xlim = c(-140,100), ylim = c(-50, 50),
#               gridlines.major = FALSE,
#               gridlines.minor = FALSE,
#               borderWidth = 1,
#               legendPosition = 'top', legendLabSize = 16, legendIconSize = 6.0) + 
#               theme(axis.text.x       = element_text(size=16,color="black"),
#                     axis.text.y       = element_text(size=16,color="black"),
#                     axis.ticks.x      = element_line(color="black"),
#                     axis.ticks.y      = element_line(color="black"))
# bp1
# 
# bp2 <- pairsplot(p,
#                 components = getComponents(p, c(1:5)),
#                 triangle = TRUE, trianglelabSize = 12,
#                 hline = 0, vline = 0,
#                 pointSize = 1,
#                 gridlines.major = FALSE, gridlines.minor = FALSE,
#                 colby = 'condition',
#                 title = 'Pairs plot', plotaxes = FALSE,
#                 margingaps = unit(c(-0.01, -0.01, -0.01, -0.01), 'cm'))
# bp2
# 
# # Plot pheatmap
# 
# rld <- rlog(dds, blind=T)
# rld_mat <- assay(rld)
# pca <- prcomp(t(rld_mat))
# rld_cor <- cor(rld_mat)
# bp3 <- pheatmap(rld_cor, annotation = coldata_trt[,1:2])
# 
# bp3
# 
# setwd("..")
# setwd("output/diploid_heat")
# ggexport(filename = "DIPLOID-HEAT-PCA.png",
#          plot   = bp1,
#          res    = 600,
#          # device = "png",
#          width  = 4000,
#          height = 5000)
# 
# ggexport(filename = "DIPLOID-HEAT-PAIRS.png",
#          plot   = bp2,
#          res    = 600,
#          width  = 6000,
#          height = 6000)
# 
# ggexport(filename = "DIPLOID-HEAT-pheatmap.png",
#          plot   = bp3,
#          res    = 600,
#          width  = 6000,
#          height = 6000)
# setwd("..")

# pcaExplorer(dds = dds_ploidy)


```

# DIPLOID only - effect of desiccation
```{r DIPLOID:DESICCATION, warning=FALSE, include=TRUE}

# A - diploid; control
# B - triploid; control
# C - diploid; heat
# D - triploid; heat
# E - diploid; desiccation
# F - triploid; desiccation

# Filter data
coldata_trt <- coldata %>% filter(group == "A" | group == "E")
cts_trt     <- subset(cts, select=row.names(coldata_trt))

# Calculate DESeq object
dds <- DESeqDataSetFromMatrix(countData = cts_trt,
                              colData = coldata_trt,
                              design = ~ condition)
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients

# Filtering: keep genes that have at least 10 counts across all samples - https://support.bioconductor.org/p/110307/
dds <- dds[rowSums(DESeq2::counts(dds)) >= 10,]

# Generate Contrasts
contrast_list    <- c("condition", "desiccation", "control") # factor, treatment group, control
res_table        <- results(dds, contrast=contrast_list, alpha = 0.05)
res_table_normal <- lfcShrink(dds,
                              coef=2, 
                              type="normal") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_apeglm <- lfcShrink(dds,
                              coef=2, 
                              type="apeglm") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_ashr   <- lfcShrink(dds,
                              coef=2, 
                              type="ashr")

# Plot MA
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)

pdf(file= "../output/diploid_desiccation/MA_plots.pdf" )
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
dev.off()

# Set cut.offs
padj.cutoff <- 0.05 # pvalues not produced with apeglm shrinkage estimator
svalue.cutoff <- 0.005 # 0.005 corresponds to pvalue of 0.05
lfc.cutoff <- 1.5 # fold-change cutoff, 1.5 or 2-fold is recommended

# Rearrange tables by gene
res_table <- res_table %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_normal <- res_table_normal %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_apeglm <- res_table_apeglm %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_ashr <- res_table_ashr %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

# How many genes are differentially expressed compared to Control?
colnames(LOCID_Roslin_gigas_gene_mito_table) <- c("gene","description","feature")
all_genes           <- left_join(res_table, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes           <- all_genes %>% arrange(-log2FoldChange)
all_genes$treatment <- "diploid_desiccation"
# all_genes_pvalue    <- cbind(gene=all_genes$gene,padj=all_genes$padj)

all_genes_normal           <- left_join(res_table_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_normal           <- all_genes_normal %>% arrange(-log2FoldChange)
all_genes_normal$treatment <- "diploid_desiccation"

all_genes_apeglm           <- left_join(res_table_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_apeglm           <- all_genes_apeglm %>% arrange(-log2FoldChange)
# all_genes_apeglm           <- left_join(all_genes_apeglm, as.data.frame(all_genes_pvalue), by = "gene")
all_genes_apeglm$treatment <- "diploid_desiccation"

all_genes_ashr           <- left_join(res_table_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_ashr           <- all_genes_ashr %>% arrange(-log2FoldChange)
all_genes_ashr$treatment <- "diploid_desiccation"

# Filter results by
sig_genes <- res_table %>%
  filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes <- left_join(sig_genes, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes <- sig_genes %>% arrange(-log2FoldChange)
sig_genes$treatment <- "diploid_desiccation" 

sig_genes_normal <- res_table_normal %>%
  filter(abs(log2FoldChange) > lfc.cutoff)
sig_genes_normal <- left_join(sig_genes_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_normal <- sig_genes_normal %>% arrange(-log2FoldChange)
sig_genes_normal$treatment <- "diploid_desiccation" 

sig_genes_apeglm <- res_table_apeglm %>%
  filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes_apeglm <- left_join(sig_genes_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_apeglm <- sig_genes_apeglm %>% arrange(-log2FoldChange)
sig_genes_apeglm$treatment <- "diploid_desiccation" 

sig_genes_ashr <- res_table_ashr %>%
  filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes_ashr <- left_join(sig_genes_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_ashr <- sig_genes_ashr %>% arrange(-log2FoldChange)
sig_genes_ashr$treatment <- "diploid_desiccation" 

# Save single gene expression counts to a data frame object, for each shrinkage estimator
single_gene_expression_unshrunk <- data.frame()
for (i in 1:nrow(sig_genes)){
  LOCID <- sig_genes$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_unshrunk <- rbind(single_gene_expression_unshrunk,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_normal <- data.frame()
for (i in 1:nrow(sig_genes_normal)){
  LOCID <- sig_genes_normal$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_normal <- rbind(single_gene_expression_normal,gene_counts_data)
  rm(gene_counts_data)
}


single_gene_expression_apeglm <- data.frame()
for (i in 1:nrow(sig_genes_apeglm)){
  LOCID <- sig_genes_apeglm$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_apeglm <- rbind(single_gene_expression_apeglm,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_ashr <- data.frame()
for (i in 1:nrow(sig_genes_ashr)){
  LOCID <- sig_genes_ashr$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_ashr <- rbind(single_gene_expression_ashr,gene_counts_data)
  rm(gene_counts_data)
}

write.table(single_gene_expression_unshrunk, file = "../output/diploid_desiccation/DIPLOID_DESICCATION-unshrunk-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_normal, file = "../output/diploid_desiccation/DIPLOID_DESICCATION-normal-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_apeglm, file = "../output/diploid_desiccation/DIPLOID_DESICCATION-apeglm-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_ashr, file = "../output/diploid_desiccation/DIPLOID_DESICCATION-ashr-single-gene-counts.csv", row.names = FALSE)

print(paste("# genes before filtering:", nrow(cts))) # count number of genes before filtering
print(paste("# genes after filtering:", length(dds))) # count number of genes after filtering
print(paste("# of genes dropped:", nrow(cts) - length(dds), sep=" ")) # count number of genes dropped
print(paste("# DEGs, all-genes:", nrow(all_genes)))
print(paste("# DEGs, unshrunken, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes)))
print(paste("# DEGs, normal shrinkage estimator, lfc = 1.5:", nrow(sig_genes_normal)))
print(paste("# DEGs, apeglm shrinkage estimator, svalue = 0.005, lfc = 1.5:", nrow(sig_genes_apeglm)))
print(paste("# DEGs, ashr shrinkage estimator, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes_ashr)))
gene_counts <- as.data.frame(c(nrow(cts),length(dds),(nrow(cts)-length(dds)),nrow(all_genes),
                            nrow(all_genes_normal),nrow(all_genes_apeglm),nrow(all_genes_ashr),
                            nrow(sig_genes),nrow(sig_genes_normal),
                            nrow(sig_genes_apeglm),nrow(sig_genes_ashr)))
row.names(gene_counts) <- c("genes_before_filtering","genes_after_filtering","genes_dropped",
                            "DEGs_all-genes","DEGs_all-genes-normal",
                            "DEGs_all-genes-apeglm","DEGs_all-genes-ashr",
                            "DEG_unshrunken-p0.05_lfc1.5","DEG_normal-lfc1.5",
                            "DEG_apeglm-s0.005_lfc1.5","DEG_ashr-p0.05_lfc1.5")
colnames(gene_counts) <- "count"

# Output tables of DEGs
write.table(all_genes,        file = "../output/diploid_desiccation/DIPLOID_DESICCATION-ALL-DEG.csv",            row.names = FALSE)
write.table(all_genes_normal, file = "../output/diploid_desiccation/DIPLOID_DESICCATION-ALL-DEG-normal.csv",     row.names = FALSE)
write.table(all_genes_apeglm, file = "../output/diploid_desiccation/DIPLOID_DESICCATION-ALL-DEG-apeglm.csv",     row.names = FALSE)
write.table(all_genes_ashr,   file = "../output/diploid_desiccation/DIPLOID_DESICCATION-ALL-DEG-ashr.csv",       row.names = FALSE)
write.table(sig_genes,        file = "../output/diploid_desiccation/DIPLOID_DESICCATION-SIG-DEG-unshrunken.csv", row.names = FALSE)
write.table(sig_genes_normal, file = "../output/diploid_desiccation/DIPLOID_DESICCATION-SIG-DEG-normal.csv",     row.names = FALSE)
write.table(sig_genes_apeglm, file = "../output/diploid_desiccation/DIPLOID_DESICCATION-SIG-DEG-apeglm.csv",     row.names = FALSE)
write.table(sig_genes_ashr,   file = "../output/diploid_desiccation/DIPLOID_DESICCATION-SIG-DEG-ashr.csv",       row.names = FALSE)
write.table(gene_counts,      file = "../output/diploid_desiccation/DIPLOID_DESICCATION-gene-counts.csv",        row.names = TRUE)

# # Plot PCA
# p <- pca(assay(vst(dds)), metadata = coldata_trt)
# screeplot(p, axisLabSize = 18, titleLabSize = 22)
# biplot(p, showLoadings = TRUE, colby = 'group',
#     labSize = 5, pointSize = 5, sizeLoadingsNames = 5)
# 
# bp1 <- biplot(p, x = "PC1", y = "PC2",
#               colby = 'group',
#               colkey = c('A' = 'royalblue1' , 
#                          'B' = 'green1'     ,
#                          'C' = 'yellow2' ,
#                          'D' = 'orange' ,
#                          'E' = 'orangered1',
#                          'F' = 'red3'),
#               shape = 'condition',
#               shapekey = c('control' = 19, 'desiccation' = 19),
#               pointSize = 4,
#               showLoadings = FALSE,
#               lab = NULL,
#               drawConnectors = FALSE,
#               ellipse = TRUE,
#               ellipseLevel = 0.95,
#               ellipseFill = FALSE,
#               ellipseFillKey = c('control' = 'royalblue1', 'desiccation' = 'red1'),
#               ellipseAlpha = 1/4,
#               ellipseLineSize = 1,
#               xlim = c(-150,180), ylim = c(-110, 60),
#               gridlines.major = FALSE,
#               gridlines.minor = FALSE,
#               borderWidth = 1,
#               legendPosition = 'top', legendLabSize = 16, legendIconSize = 6.0) + 
#               theme(axis.text.x       = element_text(size=16,color="black"),
#                     axis.text.y       = element_text(size=16,color="black"),
#                     axis.ticks.x      = element_line(color="black"),
#                     axis.ticks.y      = element_line(color="black"))
# bp1
# 
# bp2 <- pairsplot(p,
#                 components = getComponents(p, c(1:5)),
#                 triangle = TRUE, trianglelabSize = 12,
#                 hline = 0, vline = 0,
#                 pointSize = 1,
#                 gridlines.major = FALSE, gridlines.minor = FALSE,
#                 colby = 'condition',
#                 title = 'Pairs plot', plotaxes = FALSE,
#                 margingaps = unit(c(-0.01, -0.01, -0.01, -0.01), 'cm'))
# bp2
# 
# 
# # Plot pheatmap
# rld <- rlog(dds, blind=T)
# rld_mat <- assay(rld)
# pca <- prcomp(t(rld_mat))
# rld_cor <- cor(rld_mat)
# bp3 <- pheatmap(rld_cor, annotation = coldata_trt[,1:2])
# 
# bp3
# 
# setwd("..")
# setwd("output/diploid_desiccation")
# ggexport(filename = "DIPLOID-DESICCATION-PCA.png",
#          plot   = bp1,
#          res    = 600,
#          # device = "png",
#          width  = 4000,
#          height = 5000)
# 
# ggexport(filename = "DIPLOID-DESICCATION-PAIRS.png",
#          plot   = bp2,
#          res    = 600,
#          width  = 6000,
#          height = 6000)
# 
# ggexport(filename = "DIPLOID-DESICCATION-pheatmap.png",
#          plot   = bp3,
#          res    = 600,
#          width  = 6000,
#          height = 6000)
# setwd("..")

# pcaExplorer(dds = dds_ploidy)


```

# TRIPLOID only - effect of temperature
```{r TRIPLOID:HEAT, warning=FALSE, include=TRUE}

# A - diploid; control
# B - triploid; control
# C - diploid; heat
# D - triploid; heat
# E - diploid; desiccation
# F - triploid; desiccation

# Filter data
coldata_trt <- coldata %>% filter(group == "B" | group == "D")
cts_trt     <- subset(cts, select=row.names(coldata_trt))

# Calculate DESeq object
dds <- DESeqDataSetFromMatrix(countData = cts_trt,
                              colData = coldata_trt,
                              design = ~ condition)
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients

# Filtering: keep genes that have at least 10 counts across all samples - https://support.bioconductor.org/p/110307/
dds <- dds[rowSums(DESeq2::counts(dds)) >= 10,]

# Generate Contrasts
contrast_list    <- c("condition", "heat", "control") # factor, treatment group, control
res_table        <- results(dds, contrast=contrast_list, alpha = 0.05)
res_table_normal <- lfcShrink(dds,
                              coef=2, 
                              type="normal") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_apeglm <- lfcShrink(dds,
                              coef=2, 
                              type="apeglm") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_ashr   <- lfcShrink(dds,
                              coef=2, 
                              type="ashr")

# Plot MA
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)

pdf(file= "../output/triploid_heat/MA_plots.pdf" )
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
dev.off()

# Set cut.offs
padj.cutoff <- 0.05 # pvalues not produced with apeglm shrinkage estimator
svalue.cutoff <- 0.005 # 0.005 corresponds to pvalue of 0.05
lfc.cutoff <- 1.5 # fold-change cutoff, 1.5 or 2-fold is recommended

# Rearrange tables by gene
res_table <- res_table %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_normal <- res_table_normal %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_apeglm <- res_table_apeglm %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_ashr <- res_table_ashr %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

# How many genes are differentially expressed compared to Control?
colnames(LOCID_Roslin_gigas_gene_mito_table) <- c("gene","description","feature")
all_genes           <- left_join(res_table, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes           <- all_genes %>% arrange(-log2FoldChange)
all_genes$treatment <- "triploid_heat"
# all_genes_pvalue    <- cbind(gene=all_genes$gene,padj=all_genes$padj)

all_genes_normal           <- left_join(res_table_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_normal           <- all_genes_normal %>% arrange(-log2FoldChange)
all_genes_normal$treatment <- "triploid_heat"

all_genes_apeglm           <- left_join(res_table_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_apeglm           <- all_genes_apeglm %>% arrange(-log2FoldChange)
# all_genes_apeglm           <- left_join(all_genes_apeglm, as.data.frame(all_genes_pvalue), by = "gene")
all_genes_apeglm$treatment <- "triploid_heat"

all_genes_ashr           <- left_join(res_table_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_ashr           <- all_genes_ashr %>% arrange(-log2FoldChange)
all_genes_ashr$treatment <- "triploid_heat"

# Filter results by
sig_genes <- res_table %>%
  filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes <- left_join(sig_genes, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes <- sig_genes %>% arrange(-log2FoldChange)
sig_genes$treatment <- "triploid_heat" 

sig_genes_normal <- res_table_normal %>%
  filter(abs(log2FoldChange) > lfc.cutoff)
sig_genes_normal <- left_join(sig_genes_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_normal <- sig_genes_normal %>% arrange(-log2FoldChange)
sig_genes_normal$treatment <- "triploid_heat" 

sig_genes_apeglm <- res_table_apeglm %>%
  filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes_apeglm <- left_join(sig_genes_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_apeglm <- sig_genes_apeglm %>% arrange(-log2FoldChange)
sig_genes_apeglm$treatment <- "triploid_heat" 

sig_genes_ashr <- res_table_ashr %>%
  filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes_ashr <- left_join(sig_genes_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_ashr <- sig_genes_ashr %>% arrange(-log2FoldChange)
sig_genes_ashr$treatment <- "triploid_heat" 

# Save single gene expression counts to a data frame object, for each shrinkage estimator
single_gene_expression_unshrunk <- data.frame()
for (i in 1:nrow(sig_genes)){
  LOCID <- sig_genes$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_unshrunk <- rbind(single_gene_expression_unshrunk,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_normal <- data.frame()
for (i in 1:nrow(sig_genes_normal)){
  LOCID <- sig_genes_normal$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_normal <- rbind(single_gene_expression_normal,gene_counts_data)
  rm(gene_counts_data)
}


single_gene_expression_apeglm <- data.frame()
for (i in 1:nrow(sig_genes_apeglm)){
  LOCID <- sig_genes_apeglm$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_apeglm <- rbind(single_gene_expression_apeglm,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_ashr <- data.frame()
for (i in 1:nrow(sig_genes_ashr)){
  LOCID <- sig_genes_ashr$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_ashr <- rbind(single_gene_expression_ashr,gene_counts_data)
  rm(gene_counts_data)
}

write.table(single_gene_expression_unshrunk, file = "../output/triploid_heat/TRIPLOID_HEAT-unshrunk-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_normal, file = "../output/triploid_heat/TRIPLOID_HEAT-normal-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_apeglm, file = "../output/triploid_heat/TRIPLOID_HEAT-apeglm-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_ashr, file = "../output/triploid_heat/TRIPLOID_HEAT-ashr-single-gene-counts.csv", row.names = FALSE)

print(paste("# genes before filtering:", nrow(cts))) # count number of genes before filtering
print(paste("# genes after filtering:", length(dds))) # count number of genes after filtering
print(paste("# of genes dropped:", nrow(cts) - length(dds), sep=" ")) # count number of genes dropped
print(paste("# DEGs, all-genes:", nrow(all_genes)))
print(paste("# DEGs, unshrunken, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes)))
print(paste("# DEGs, normal shrinkage estimator, lfc = 1.5:", nrow(sig_genes_normal)))
print(paste("# DEGs, apeglm shrinkage estimator, svalue = 0.005, lfc = 1.5:", nrow(sig_genes_apeglm)))
print(paste("# DEGs, ashr shrinkage estimator, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes_ashr)))
gene_counts <- as.data.frame(c(nrow(cts),length(dds),(nrow(cts)-length(dds)),nrow(all_genes),
                            nrow(all_genes_normal),nrow(all_genes_apeglm),nrow(all_genes_ashr),
                            nrow(sig_genes),nrow(sig_genes_normal),
                            nrow(sig_genes_apeglm),nrow(sig_genes_ashr)))
row.names(gene_counts) <- c("genes_before_filtering","genes_after_filtering","genes_dropped",
                            "DEGs_all-genes","DEGs_all-genes-normal",
                            "DEGs_all-genes-apeglm","DEGs_all-genes-ashr",
                            "DEG_unshrunken-p0.05_lfc1.5","DEG_normal-lfc1.5",
                            "DEG_apeglm-s0.005_lfc1.5","DEG_ashr-p0.05_lfc1.5")
colnames(gene_counts) <- "count"

# Output tables of DEGs
write.table(all_genes,        file = "../output/triploid_heat/TRIPLOID_HEAT-ALL-DEG.csv",            row.names = FALSE)
write.table(all_genes_normal, file = "../output/triploid_heat/TRIPLOID_HEAT-ALL-DEG-normal.csv",     row.names = FALSE)
write.table(all_genes_apeglm, file = "../output/triploid_heat/TRIPLOID_HEAT-ALL-DEG-apeglm.csv",     row.names = FALSE)
write.table(all_genes_ashr,   file = "../output/triploid_heat/TRIPLOID_HEAT-ALL-DEG-ashr.csv",       row.names = FALSE)
write.table(sig_genes,        file = "../output/triploid_heat/TRIPLOID_HEAT-SIG-DEG-unshrunken.csv", row.names = FALSE)
write.table(sig_genes_normal, file = "../output/triploid_heat/TRIPLOID_HEAT-SIG-DEG-normal.csv",     row.names = FALSE)
write.table(sig_genes_apeglm, file = "../output/triploid_heat/TRIPLOID_HEAT-SIG-DEG-apeglm.csv",     row.names = FALSE)
write.table(sig_genes_ashr,   file = "../output/triploid_heat/TRIPLOID_HEAT-SIG-DEG-ashr.csv",       row.names = FALSE)
write.table(gene_counts,      file = "../output/triploid_heat/TRIPLOID_HEAT-gene-counts.csv",        row.names = TRUE)

# # Plot PCA
# p <- pca(assay(vst(dds)), metadata = coldata_trt)
# screeplot(p, axisLabSize = 18, titleLabSize = 22)
# biplot(p, showLoadings = TRUE, colby = 'group',
#     labSize = 5, pointSize = 5, sizeLoadingsNames = 5)
# 
# bp1 <- biplot(p, x = "PC1", y = "PC2",
#               colby = 'group',
#               colkey = c('A' = 'royalblue1' , 
#                          'B' = 'green1'     ,
#                          'C' = 'yellow2' ,
#                          'D' = 'orange' ,
#                          'E' = 'orangered1',
#                          'F' = 'red3'),
#               shape = 'condition',
#               shapekey = c('control' = 17, 'heat' = 17),
#               pointSize = 4,
#               showLoadings = FALSE,
#               lab = NULL,
#               drawConnectors = FALSE,
#               ellipse = TRUE,
#               ellipseLevel = 0.95,
#               ellipseFill = FALSE,
#               ellipseFillKey = c('control' = 'green1', 'heat' = 'orangered3'),
#               ellipseAlpha = 1/4,
#               ellipseLineSize = 1,
#               xlim = c(-160,140), ylim = c(-100, 80),
#               gridlines.major = FALSE,
#               gridlines.minor = FALSE,
#               borderWidth = 1,
#               legendPosition = 'top', legendLabSize = 16, legendIconSize = 6.0) + 
#               theme(axis.text.x       = element_text(size=16,color="black"),
#                     axis.text.y       = element_text(size=16,color="black"),
#                     axis.ticks.x      = element_line(color="black"),
#                     axis.ticks.y      = element_line(color="black"))
# bp1
# 
# bp2 <- pairsplot(p,
#                 components = getComponents(p, c(1:5)),
#                 triangle = TRUE, trianglelabSize = 12,
#                 hline = 0, vline = 0,
#                 pointSize = 1,
#                 gridlines.major = FALSE, gridlines.minor = FALSE,
#                 colby = 'condition',
#                 title = 'Pairs plot', plotaxes = FALSE,
#                 margingaps = unit(c(-0.01, -0.01, -0.01, -0.01), 'cm'))
# bp2
# 
# # Plot pheatmap
# rld <- rlog(dds, blind=T)
# rld_mat <- assay(rld)
# pca <- prcomp(t(rld_mat))
# rld_cor <- cor(rld_mat)
# bp3 <- pheatmap(rld_cor, annotation = coldata_trt[,1:2])
# 
# bp3
# 
# setwd("..")
# setwd("output/triploid_heat")
# ggexport(filename = "TRIPLOID-HEAT-PCA.png",
#          plot   = bp1,
#          res    = 600,
#          # device = "png",
#          width  = 4000,
#          height = 5000)
# 
# ggexport(filename = "TRIPLOID-HEAT-PAIRS.png",
#          plot   = bp2,
#          res    = 600,
#          width  = 6000,
#          height = 6000)
# 
# ggexport(filename = "TRIPLOID-HEAT-pheatmap.png",
#          plot   = bp3,
#          res    = 600,
#          width  = 6000,
#          height = 6000)
# setwd("..")

# pcaExplorer(dds = dds_ploidy)


```

# TRIPLOID only - effect of desiccation
```{r TRIPLOID:DESICCATION, warning=FALSE, include=TRUE}

# A - diploid; control
# B - triploid; control
# C - diploid; heat
# D - triploid; heat
# E - diploid; desiccation
# F - triploid; desiccation

# Filter data
coldata_trt <- coldata %>% filter(group == "B" | group == "F")
cts_trt     <- subset(cts, select=row.names(coldata_trt))

# Calculate DESeq object
dds <- DESeqDataSetFromMatrix(countData = cts_trt,
                              colData = coldata_trt,
                              design = ~ condition)
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients

# Filtering: keep genes that have at least 10 counts across all samples - https://support.bioconductor.org/p/110307/
dds <- dds[rowSums(DESeq2::counts(dds)) >= 10,]

# Generate Contrasts
contrast_list    <- c("condition", "desiccation", "control") # factor, treatment group, control
res_table        <- results(dds, contrast=contrast_list, alpha = 0.05)
res_table_normal <- lfcShrink(dds,
                              coef=2, 
                              type="normal") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_apeglm <- lfcShrink(dds,
                              coef=2, 
                              type="apeglm") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_ashr   <- lfcShrink(dds,
                              coef=2, 
                              type="ashr")

# Plot MA
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)

pdf(file= "../output/triploid_desiccation/MA_plots.pdf" )
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
dev.off()

# Set cut.offs
padj.cutoff <- 0.05 # pvalues not produced with apeglm shrinkage estimator
svalue.cutoff <- 0.005 # 0.005 corresponds to pvalue of 0.05
lfc.cutoff <- 1.5 # fold-change cutoff, 1.5 or 2-fold is recommended

# Rearrange tables by gene
res_table <- res_table %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_normal <- res_table_normal %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_apeglm <- res_table_apeglm %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_ashr <- res_table_ashr %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

# How many genes are differentially expressed compared to Control?
colnames(LOCID_Roslin_gigas_gene_mito_table) <- c("gene","description","feature")
all_genes           <- left_join(res_table, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes           <- all_genes %>% arrange(-log2FoldChange)
all_genes$treatment <- "triploid_desiccation"
# all_genes_pvalue    <- cbind(gene=all_genes$gene,padj=all_genes$padj)

all_genes_normal           <- left_join(res_table_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_normal           <- all_genes_normal %>% arrange(-log2FoldChange)
all_genes_normal$treatment <- "triploid_desiccation"

all_genes_apeglm           <- left_join(res_table_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_apeglm           <- all_genes_apeglm %>% arrange(-log2FoldChange)
# all_genes_apeglm           <- left_join(all_genes_apeglm, as.data.frame(all_genes_pvalue), by = "gene")
all_genes_apeglm$treatment <- "triploid_desiccation"

all_genes_ashr           <- left_join(res_table_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_ashr           <- all_genes_ashr %>% arrange(-log2FoldChange)
all_genes_ashr$treatment <- "triploid_desiccation"

# Filter results by
sig_genes <- res_table %>%
  filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes <- left_join(sig_genes, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes <- sig_genes %>% arrange(-log2FoldChange)
sig_genes$treatment <- "triploid_desiccation" 

sig_genes_normal <- res_table_normal %>%
  filter(abs(log2FoldChange) > lfc.cutoff)
sig_genes_normal <- left_join(sig_genes_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_normal <- sig_genes_normal %>% arrange(-log2FoldChange)
sig_genes_normal$treatment <- "triploid_desiccation" 

sig_genes_apeglm <- res_table_apeglm %>%
  filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes_apeglm <- left_join(sig_genes_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_apeglm <- sig_genes_apeglm %>% arrange(-log2FoldChange)
sig_genes_apeglm$treatment <- "triploid_desiccation" 

sig_genes_ashr <- res_table_ashr %>%
  filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes_ashr <- left_join(sig_genes_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_ashr <- sig_genes_ashr %>% arrange(-log2FoldChange)
sig_genes_ashr$treatment <- "triploid_desiccation" 

# Save single gene expression counts to a data frame object, for each shrinkage estimator
single_gene_expression_unshrunk <- data.frame()
for (i in 1:nrow(sig_genes)){
  LOCID <- sig_genes$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_unshrunk <- rbind(single_gene_expression_unshrunk,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_normal <- data.frame()
for (i in 1:nrow(sig_genes_normal)){
  LOCID <- sig_genes_normal$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_normal <- rbind(single_gene_expression_normal,gene_counts_data)
  rm(gene_counts_data)
}


single_gene_expression_apeglm <- data.frame()
for (i in 1:nrow(sig_genes_apeglm)){
  LOCID <- sig_genes_apeglm$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_apeglm <- rbind(single_gene_expression_apeglm,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_ashr <- data.frame()
for (i in 1:nrow(sig_genes_ashr)){
  LOCID <- sig_genes_ashr$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_ashr <- rbind(single_gene_expression_ashr,gene_counts_data)
  rm(gene_counts_data)
}

write.table(single_gene_expression_unshrunk, file = "../output/triploid_desiccation/TRIPLOID_DESICCATION-unshrunk-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_normal, file = "../output/triploid_desiccation/TRIPLOID_DESICCATION-normal-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_apeglm, file = "../output/triploid_desiccation/TRIPLOID_DESICCATION-apeglm-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_ashr, file = "../output/triploid_desiccation/TRIPLOID_DESICCATION-ashr-single-gene-counts.csv", row.names = FALSE)

print(paste("# genes before filtering:", nrow(cts))) # count number of genes before filtering
print(paste("# genes after filtering:", length(dds))) # count number of genes after filtering
print(paste("# of genes dropped:", nrow(cts) - length(dds), sep=" ")) # count number of genes dropped
print(paste("# DEGs, all-genes:", nrow(all_genes)))
print(paste("# DEGs, unshrunken, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes)))
print(paste("# DEGs, normal shrinkage estimator, lfc = 1.5:", nrow(sig_genes_normal)))
print(paste("# DEGs, apeglm shrinkage estimator, svalue = 0.005, lfc = 1.5:", nrow(sig_genes_apeglm)))
print(paste("# DEGs, ashr shrinkage estimator, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes_ashr)))
gene_counts <- as.data.frame(c(nrow(cts),length(dds),(nrow(cts)-length(dds)),nrow(all_genes),
                            nrow(all_genes_normal),nrow(all_genes_apeglm),nrow(all_genes_ashr),
                            nrow(sig_genes),nrow(sig_genes_normal),
                            nrow(sig_genes_apeglm),nrow(sig_genes_ashr)))
row.names(gene_counts) <- c("genes_before_filtering","genes_after_filtering","genes_dropped",
                            "DEGs_all-genes","DEGs_all-genes-normal",
                            "DEGs_all-genes-apeglm","DEGs_all-genes-ashr",
                            "DEG_unshrunken-p0.05_lfc1.5","DEG_normal-lfc1.5",
                            "DEG_apeglm-s0.005_lfc1.5","DEG_ashr-p0.05_lfc1.5")
colnames(gene_counts) <- "count"

# Output tables of DEGs
write.table(all_genes,        file = "../output/triploid_desiccation/TRIPLOID_DESICCATION-ALL-DEG.csv",            row.names = FALSE)
write.table(all_genes_normal, file = "../output/triploid_desiccation/TRIPLOID_DESICCATION-ALL-DEG-normal.csv",     row.names = FALSE)
write.table(all_genes_apeglm, file = "../output/triploid_desiccation/TRIPLOID_DESICCATION-ALL-DEG-apeglm.csv",     row.names = FALSE)
write.table(all_genes_ashr,   file = "../output/triploid_desiccation/TRIPLOID_DESICCATION-ALL-DEG-ashr.csv",       row.names = FALSE)
write.table(sig_genes,        file = "../output/triploid_desiccation/TRIPLOID_DESICCATION-SIG-DEG-unshrunken.csv", row.names = FALSE)
write.table(sig_genes_normal, file = "../output/triploid_desiccation/TRIPLOID_DESICCATION-SIG-DEG-normal.csv",     row.names = FALSE)
write.table(sig_genes_apeglm, file = "../output/triploid_desiccation/TRIPLOID_DESICCATION-SIG-DEG-apeglm.csv",     row.names = FALSE)
write.table(sig_genes_ashr,   file = "../output/triploid_desiccation/TRIPLOID_DESICCATION-SIG-DEG-ashr.csv",       row.names = FALSE)
write.table(gene_counts,      file = "../output/triploid_desiccation/TRIPLOID_DESICCATION-gene-counts.csv",        row.names = TRUE)

# # Plot PCA
# p <- pca(assay(vst(dds)), metadata = coldata_trt)
# screeplot(p, axisLabSize = 18, titleLabSize = 22)
# biplot(p, showLoadings = TRUE, colby = 'group',
#     labSize = 5, pointSize = 5, sizeLoadingsNames = 5)
# 
# bp1 <- biplot(p, x = "PC1", y = "PC2",
#               colby = 'group',
#               colkey = c('A' = 'royalblue1' , 
#                          'B' = 'green1'     ,
#                          'C' = 'yellow2' ,
#                          'D' = 'orange' ,
#                          'E' = 'orangered1',
#                          'F' = 'red3'),
#               shape = 'condition',
#               shapekey = c('control' = 17, 'desiccation' = 17),
#               pointSize = 4,
#               showLoadings = FALSE,
#               lab = NULL,
#               drawConnectors = FALSE,
#               ellipse = TRUE,
#               ellipseLevel = 0.95,
#               ellipseFill = FALSE,
#               ellipseFillKey = c('control' = 'green1', 'desiccation' = 'red3'),
#               ellipseAlpha = 1/4,
#               ellipseLineSize = 1,
#               xlim = c(-150,130), ylim = c(-70, 60),
#               gridlines.major = FALSE,
#               gridlines.minor = FALSE,
#               borderWidth = 1,
#               legendPosition = 'top', legendLabSize = 16, legendIconSize = 6.0) + 
#               theme(axis.text.x       = element_text(size=16,color="black"),
#                     axis.text.y       = element_text(size=16,color="black"),
#                     axis.ticks.x      = element_line(color="black"),
#                     axis.ticks.y      = element_line(color="black"))
# bp1
# 
# 
# bp2 <- pairsplot(p,
#                 components = getComponents(p, c(1:5)),
#                 triangle = TRUE, trianglelabSize = 12,
#                 hline = 0, vline = 0,
#                 pointSize = 1,
#                 gridlines.major = FALSE, gridlines.minor = FALSE,
#                 colby = 'condition',
#                 title = 'Pairs plot', plotaxes = FALSE,
#                 margingaps = unit(c(-0.01, -0.01, -0.01, -0.01), 'cm'))
# bp2
# 
# # Plot pheatmap
# rld <- rlog(dds, blind=T)
# rld_mat <- assay(rld)
# pca <- prcomp(t(rld_mat))
# rld_cor <- cor(rld_mat)
# bp3 <- pheatmap(rld_cor, annotation = coldata_trt[,1:2])
# 
# bp3
# 
# setwd("..")
# setwd("output/triploid_desiccation")
# ggexport(filename = "TRIPLOID-DESICCATION-PCA.png",
#          plot   = bp1,
#          res    = 600,
#          width  = 4000,
#          height = 5000)
# 
# ggexport(filename = "TRIPLOID-DESICCATION-PAIRS.png",
#          plot   = bp2,
#          res    = 600,
#          width  = 6000,
#          height = 6000)
# 
# ggexport(filename = "TRIPLOID-DESICCATION-pheatmap.png",
#          plot   = bp3,
#          res    = 600,
#          width  = 6000,
#          height = 6000)
# setwd("..")

# pcaExplorer(dds = dds_ploidy)

```

# Generate table for IPA
```{R}
rm(list=ls())

control_ploidy     <- read.table("../output/control_ploidy/CONTROL_PLOIDY-ALL-DEG-apeglm.csv", header = TRUE)
heat_ploidy        <- read.table("../output/heat_ploidy/HEAT_PLOIDY-ALL-DEG-apeglm.csv", header = TRUE)
desiccation_ploidy <- read.table("../output/desiccation_ploidy/DESICCATION_PLOIDY-ALL-DEG-apeglm.csv", header = TRUE)
diploid_heat          <- read.table("../output/diploid_heat/DIPLOID_HEAT-ALL-DEG-apeglm.csv", header = TRUE)
diploid_desciccation  <- read.table("../output/diploid_desiccation/DIPLOID_DESICCATION-ALL-DEG-apeglm.csv", header = TRUE)
triploid_heat         <- read.table("../output/triploid_heat/TRIPLOID_HEAT-ALL-DEG-apeglm.csv", header = TRUE)
triploid_desiccation  <- read.table("../output/triploid_desiccation/TRIPLOID_DESICCATION-ALL-DEG-apeglm.csv", header = TRUE)

control_ploidy       <- control_ploidy        %>% dplyr::select(treatment,gene,padj,log2FoldChange)
heat_ploidy          <- heat_ploidy           %>% dplyr::select(treatment,gene,padj,log2FoldChange)
desiccation_ploidy   <- desiccation_ploidy    %>% dplyr::select(treatment,gene,padj,log2FoldChange)
diploid_heat         <- diploid_heat          %>% dplyr::select(treatment,gene,padj,log2FoldChange)
diploid_desciccation <- diploid_desciccation  %>% dplyr::select(treatment,gene,padj,log2FoldChange)
triploid_heat        <- triploid_heat         %>% dplyr::select(treatment,gene,padj,log2FoldChange)
triploid_desiccation <- triploid_desiccation  %>% dplyr::select(treatment,gene,padj,log2FoldChange)

all_trt <- rbind(control_ploidy,
                 heat_ploidy,
                 desiccation_ploidy,
                 diploid_heat,
                 diploid_desciccation,
                 triploid_heat,
                 triploid_desiccation)

# check row sums match
sum(nrow(control_ploidy),nrow(heat_ploidy),nrow(desiccation_ploidy ),
    nrow(diploid_heat ),nrow(diploid_desciccation),nrow(triploid_heat ),
    nrow(triploid_desiccation))
nrow(all_trt)

write.table(all_trt, "../output/2022_cgigas_DEG-all-treatments.txt", row.names = FALSE)

```

